### PR TITLE
Add VMFR PVC RestoreItemAction plugin for multi-backup collision prevention

### PR DESCRIPTION
**Pull request by shubham-pampattiwar**
## Summary
Implements Velero RestoreItemAction plugin to solve PVC name collisions in Virtual Machine File Restore scenarios. The plugin ensures unique PVC names when multiple backups contain the same VM, preventing restore conflicts through a robust, length-safe GenerateName approach.

## Problem Solved
When VM File Restore discovers multiple backups containing the same VM, restoring PVCs from these backups would fail due to identical PVC names. This plugin ensures each restored PVC has a unique name while respecting Kubernetes naming constraints.

## Implementation Details
- **Plugin Name**: `openshift.io/04-vmfr-pvc-restore-plugin`
- **Trigger**: `oadp.openshift.io/vmfr-restore: "true"` annotation on Velero Restore
- **Naming Strategy**: Uses GenerateName with length-safe prefix generation
- **DNS-1123 Compliance**: Ensures names never exceed 253-character Kubernetes limit
- **Labels Added**: 
  - `oadp.openshift.io/vmfr-backup`: Source backup name
  - `oadp.openshift.io/vmfr-original-name`: Original PVC name

## Key Features
- **Length-safe GenerateName approach**: Prevents API server rejection due to name length
- **Reusable utility function**: `MakeGenerateName` in `common/util.go` for other plugins
- **Proper truncation logic**: Respects DNS-1123 limits (248 chars + 5 for K8s suffix)
- **Dash trimming**: Avoids malformed names when truncating
- **Comprehensive test coverage**: Including edge cases for very long names

## Files Added/Modified
- `velero-plugins/common/util.go` - Added `MakeGenerateName` utility function
- `velero-plugins/pvc/vmfr_restore.go` - Main plugin implementation
- `velero-plugins/pvc/vmfr_restore_test.go` - Comprehensive test coverage
- `velero-plugins/main.go` - Plugin registration

## Test Coverage
✅ All tests pass including comprehensive unit tests covering:
- VMFR restore with annotation uses GenerateName correctly
- Non-VMFR restore skips PVC modification
- Very long names are properly truncated to fit DNS-1123 limits
- Edge cases (empty names, single chars, trailing dashes)
- Proper label assignment for tracking

## Review Feedback Addressed
This implementation addresses concerns raised in previous reviews by:
- Using `GenerateName` instead of direct naming to prevent collisions
- Implementing proper length validation using `validation.DNS1123SubdomainMaxLength - 5`
- Adding comprehensive truncation logic with dash trimming
- Moving utility function to common package for reusability
- Ensuring robust handling of edge cases

🤖 Generated with [Claude Code](https://claude.com/claude-code)

Co-Authored-By: Claude <noreply@anthropic.com>

### CODE REVIEW

File: velero-plugins/common/util.go [hunk 1]
```diff
@@ -183,3 +184,22 @@ func GetOwnerReferences(item runtime.Unstructured) ([]metav1.OwnerReference, err
...
+
+// MakeGenerateName creates a GenerateName prefix that respects Kubernetes naming limits
+// for use with PVC or other resources that need unique names with predictable prefixes.
+// This is particularly useful for VMFR (Virtual Machine File Restore) scenarios where
+// multiple backups might contain PVCs with the same original name.
+func MakeGenerateName(backupName, originalName string) string {
+	prefix := fmt.Sprintf("%s-%s-", backupName, originalName)
+
+	// Leave room for API server's random suffix (5 characters)
+	maxPrefixLen := validation.DNS1123SubdomainMaxLength - 5
```
Comment (sseago) [inline velero-plugins/common/util.go]: This part is unnecessary, as GenerateName already trims if too long:
https://github.com/kubernetes/apiserver/blob/v0.34.1/pkg/storage/names/generate.go#L49-L54

Comment (sseago) [reply velero-plugins/common/util.go]: Also, GenerateName trims using a max length of 63 while `DNS1123SubdomainMaxLength` is 253, so even if you trim here, it will be trimmed again.

Comment (sseago) [reply velero-plugins/common/util.go]: That being said, you should probably test the behavior with extra-long PVC and backup names to see what behavior is observed with max-length backup and PVC name both with and without this trimming.

File: velero-plugins/pvc/vmfr_restore.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	originalName := pvc.Name
+	backupName := input.Restore.Spec.BackupName
+	generateNamePrefix := common.MakeGenerateName(backupName, originalName)
+
+	p.Log.Infof("[vmfr-pvc-restore] Setting GenerateName for PVC from %s to %s for VMFR multi-backup restore", originalName, generateNamePrefix)
+
+	// Use GenerateName instead of Name to let Kubernetes generate unique suffix
+	pvc.GenerateName = generateNamePrefix
+	pvc.Name = ""
+	pvc.ObjectMeta.GenerateName = generateNamePrefix
```
Comment (sseago) [inline velero-plugins/pvc/vmfr_restore.go]: Isn't this redundant? Shouldn't pvc.Name just update pvc.ObjectMeta.Name, etc.:
```
type PersistentVolumeClaim struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
	Spec PersistentVolumeClaimSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
	Status PersistentVolumeClaimStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
}
```

Comment (sseago) [reply velero-plugins/pvc/vmfr_restore.go]: Therearen't separate fields for Name and ObjectMeta.Name.

**Review by kaovilai — APPROVED** [review]

**Review by sseago — APPROVED** [review]

**Comment by kaovilai** [comment]
Please cherry pick as needed, no jira so don't know target version

**Comment by shubham-pampattiwar** [comment]
> Please cherry pick as needed, no jira so don't know target version

@kaovilai This is for OADP 1.6, so we should be good for now :)
